### PR TITLE
Initial work for C# 9 pattern-matching changes

### PR DESCRIPTION
**Pull request by gafter**
This implements
- Parts of the spec in https://github.com/dotnet/csharplang/issues/2850
  - Parenthesized Patterns (All but exhaustive testing)
  - Relational Patterns (Syntax and Local binding only)
  - Pattern Combinators (Syntax and Local binding only)
  - No work on the Open Issues section (e.g. type flow for the `and` combinator)
- The spec in https://github.com/dotnet/csharplang/issues/2925
  - All but exhaustive testing
- Not implemented: handling designators under `not` and `or`, decision tree changes, exhaustiveness
- Does not handle https://github.com/dotnet/roslyn/issues/40149
  - e.g. A pattern written `(A[])` should be capable of matching a value of type `A[]`.

### CODE REVIEW

**Comment by gafter** [comment]
@agocke @RikkiGibson Can you please review this?  I'd be happy to walk through it in person if that helps.

File: src/Compilers/CSharp/Test/Symbol/Symbols/ConversionTests.cs [hunk 1]
```diff
@@ -9,6 +9,7 @@
 using Xunit;
 using System.Collections.Generic;
 using System;
+using System.Net.Sockets;
```
Comment (RikkiGibson) [inline src/Compilers/CSharp/Test/Symbol/Symbols/ConversionTests.cs]: appears to be added by mistake

File: src/Compilers/CSharp/Portable/Compilation/CSharpSemanticModel.cs [hunk 2]
```diff
@@ -4012,7 +4012,7 @@ private static ImmutableArray<MethodSymbol> FilterOverriddenOrHiddenMethods(Immu
                         // chose for M.
                         var call = (BoundCall)boundNodeForSyntacticParent;
                         InvocationExpressionSyntax invocation = call.Syntax as InvocationExpressionSyntax;
-                        if (invocation != null && invocation.Expression.SkipParens() == boundNode.Syntax.SkipParens() && (object)call.Method != null)
+                        if (invocation != null && invocation.Expression.SkipParens() == (boundNode.Syntax as ExpressionSyntax).SkipParens() && (object)call.Method != null)
```
Comment (RikkiGibson) [inline src/Compilers/CSharp/Portable/Compilation/CSharpSemanticModel.cs]: >(boundNode.Syntax as ExpressionSyntax) [](start = 88, length = 38)

Consider making this a cast

File: src/Compilers/CSharp/Portable/Parser/LanguageParser_Patterns.cs [hunk 3]
```diff
@@ -9,6 +9,7 @@ namespace Microsoft.CodeAnalysis.CSharp.Syntax.InternalSyntax
 
     internal partial class LanguageParser : SyntaxParser
     {
+
         /// <summary>
```
Comment (RikkiGibson) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser_Patterns.cs]: nit: empty line above

File: src/Compilers/CSharp/Portable/Parser/LanguageParser_Patterns.cs [hunk 4]
```diff
@@ -442,33 +281,28 @@ private PatternSyntax ParsePatternContinued(TypeSyntax type, Precedence preceden
                     propertyPatternClause0 == null &&
                     designation0 == null &&
                     subPatterns.Count == 1 &&
-                    subPatterns[0].NameColon == null &&
-                    subPatterns.SeparatorCount == 0)
+                    subPatterns[0].NameColon == null
```
Comment (RikkiGibson) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser_Patterns.cs]: nit: the ')' should be at the end of this line

File: src/Compilers/CSharp/Portable/Operations/CSharpOperationNodes.cs [hunk 5]
```diff
@@ -28,6 +28,24 @@ internal sealed class CSharpLazyNoneOperation : LazyNoneOperation
     }
 
 
+    internal sealed class CSharpLazyNonePatternOperation : LazyNoneOperation, IPatternOperation
```
Comment (RikkiGibson) [inline src/Compilers/CSharp/Portable/Operations/CSharpOperationNodes.cs]: Is this a placeholder until the IOperation part of the feature is done? If so consider adding a PROTOTYPE comment.

File: src/Test/Utilities/Portable/TestResource.resx [hunk 6]
```diff
@@ -209,6 +209,9 @@ namespace My
             }
         }
 
+        public static bool IsAsciiLetter(char c) => c is (>= 'a' and &lt;= 'z') or (>= 'A' and &lt;= 'Z');
```
Comment (RikkiGibson) [inline src/Test/Utilities/Portable/TestResource.resx]: This appears to be only partially XML escaped. Is it only necessary to escape the '<' and not the '>'?

File: src/EditorFeatures/CSharpTest/Completion/CompletionProviders/PropertySubPatternCompletionProviderTests.cs [hunk 7]
```diff
@@ -484,7 +484,7 @@ class D
             await VerifyNoItemsExistAsync(markup);
         }
 
-        [Fact]
+        [Fact(Skip = "https://github.com/dotnet/roslyn/issues/40015")]
```
Comment (RikkiGibson) [inline src/EditorFeatures/CSharpTest/Completion/CompletionProviders/PropertySubPatternCompletionProviderTests.cs]: Should these have PROTOTYPE comments?

File: src/EditorFeatures/CSharpTest/Completion/CompletionProviders/SymbolCompletionProviderTests.cs [hunk 8]
```diff
@@ -1319,7 +1319,7 @@ public async Task PositionalPatternFirstPosition()
             await VerifyItemExistsAsync(AddUsingDirectives("using System;", AddInsideMethod(content)), @"System");
         }
 
-        [Fact, Trait(Traits.Feature, Traits.Features.Completion)]
+        [Fact(Skip = "https://github.com/dotnet/roslyn/issues/40015"), Trait(Traits.Feature, Traits.Features.Completion)]
```
Comment (RikkiGibson) [inline src/EditorFeatures/CSharpTest/Completion/CompletionProviders/SymbolCompletionProviderTests.cs]: Should this also have a prototype comment?

File: src/EditorFeatures/CSharpTest2/Recommendations/WhenKeywordRecommenderTests.cs [hunk 9]
```diff
@@ -191,15 +191,15 @@ public async Task TestForCatchClause_NotAfterFilter3()
         public async Task TestForSwitchCase_NotInEmptySwitchStatement() =>
             await VerifyAbsenceAsync(AddInsideMethod(@"switch (1) { $$ }"));
 
-        [Fact, Trait(Traits.Feature, Traits.Features.KeywordRecommending)]
+        [Fact(Skip = "https://github.com/dotnet/roslyn/issues/40015"), Trait(Traits.Feature, Traits.Features.KeywordRecommending)]
```
Comment (RikkiGibson) [inline src/EditorFeatures/CSharpTest2/Recommendations/WhenKeywordRecommenderTests.cs]: Should these also have prototype comments?

**Comment by RikkiGibson** [comment]
>             x.ToString(); // warn

nit: please remove this comment to help make things clear.

---
Refers to: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesVsPatterns.cs:136 in 6d450fb. [](commit_id = 6d450fbf858840c0916b80cca3ccd6538e6cfd6e, deletion_comment = True)

File: src/Compilers/CSharp/Portable/Binder/Binder_Patterns.cs [hunk 10]
```diff
@@ -99,47 +99,72 @@ private BoundExpression BindSwitchExpression(SwitchExpressionSyntax node, Diagno
...
                 hasErrors = true;
             }
 
-            return BindConstantPattern(node, inputType, node.Expression, hasErrors, diagnostics, out _);
+            var constantDiagnostics = DiagnosticBag.GetInstance();
+            var constantPattern = BindConstantPattern(node, inputType, expression, hasErrors, constantDiagnostics, out bool wasExpression);
+
+            // fall back to a type pattern if it fails as a constant pattern.
+            if (!wasExpression && !IsUnderscore(innerExpression) &&
+                ((CSharpParseOptions)node.SyntaxTree.Options).IsFeatureEnabled(MessageID.IDS_FeatureTypePattern))
```
Comment (RikkiGibson) [inline src/Compilers/CSharp/Portable/Binder/Binder_Patterns.cs]: We should attempt to bind a type pattern here regardless of the language version, and then issue a LangVersion diagnostic if the feature is not enabled.

**Comment by RikkiGibson** [comment]
Done with review pass (iteration 8)

File: src/EditorFeatures/CSharpTest2/Recommendations/WhenKeywordRecommenderTests.cs (hunk 9 above)
Comment (gafter) [reply src/EditorFeatures/CSharpTest2/Recommendations/WhenKeywordRecommenderTests.cs]: No, since we have an open issue to track this.

---
In reply to: [352880470](https://github.com/dotnet/roslyn/pull/39913#discussion_r352880470) [](ancestors = 352880470)

File: src/EditorFeatures/CSharpTest/Completion/CompletionProviders/SymbolCompletionProviderTests.cs (hunk 8 above)
Comment (gafter) [reply src/EditorFeatures/CSharpTest/Completion/CompletionProviders/SymbolCompletionProviderTests.cs]: Only needed when there is no tracking issue.

---
In reply to: [352880444](https://github.com/dotnet/roslyn/pull/39913#discussion_r352880444) [](ancestors = 352880444)

File: src/Test/Utilities/Portable/TestResource.resx (hunk 6 above)
Comment (gafter) [reply src/Test/Utilities/Portable/TestResource.resx]: Correct.

---
In reply to: [352880294](https://github.com/dotnet/roslyn/pull/39913#discussion_r352880294) [](ancestors = 352880294)

File: src/Compilers/CSharp/Portable/Operations/CSharpOperationNodes.cs (hunk 5 above)
Comment (gafter) [reply src/Compilers/CSharp/Portable/Operations/CSharpOperationNodes.cs]: The PROTOTYPE comments are in `CSharpOperationFactory` around line 293.

---
In reply to: [352880196](https://github.com/dotnet/roslyn/pull/39913#discussion_r352880196) [](ancestors = 352880196)

**Comment by gafter** [comment]
@RikkiGibson I think I've responded to all of your comments.  Do you have anything else?

@agocke Can you please review this?

**Comment by gafter** [comment]
@agocke Can you please review this?

**Comment by agocke** [comment]
@gafter Could you summarize what things you would expect to be changed after this PR? Which parts of the language changes are included in this PR?

File: src/Compilers/CSharp/Test/Syntax/Parsing/NullableParsingTests.cs [hunk 11]
```diff
@@ -717,15 +717,15 @@ public void DeclarationPattern_NullableType()
         public void DeclarationPattern_NullableArray()
         {
             UsingStatement("switch (e) { case T[]? t: break; }",
-                // (1,21): error CS0443: Syntax error; value expected
+                // (1,19): error CS8652: The feature 'type pattern' is currently in Preview and *unsupported*. To use Preview features, use the 'preview' language version.
```
Comment (agocke) [inline src/Compilers/CSharp/Test/Syntax/Parsing/NullableParsingTests.cs]: This seems worrisome. I thought we disambiguated `?` by saying that it wasn't part of a type pattern, and thus it would parse to `?:` in legacy cases. Should this still be a `?:`?

**Comment by gafter** [comment]
@agocke This implements
- Parts of the spec in https://github.com/dotnet/csharplang/issues/2850
  - Parenthesized Patterns (All but exhaustive testing)
  - Relational Patterns (Syntax and Local binding only)
  - Pattern Combinators (Syntax and Local binding only)
  - No work on the Open Issues section (e.g. type flow for the `and` combinator)
- The spec in https://github.com/dotnet/csharplang/issues/2925
  - All but exhaustive testing
- Does not handle https://github.com/dotnet/roslyn/issues/40149
  - A pattern written `(A[])` should be capable of matching a value of type `A[]`.

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs [hunk 12]
```diff
@@ -7654,6 +7654,110 @@ private CommonForEachStatementSyntax ParseForEachStatement(SyntaxToken awaitToke
...
+        /// Is the following set of tokens, interpreted as a type, the type <c>var</c>?
+        /// </summary>
+        private bool IsVarType()
+        {
+            if (!this.CurrentToken.IsIdentifierVar())
+            {
+                return false;
+            }
+
+            switch (this.PeekToken(1).Kind)
```
Comment (agocke) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs]: This seems like we're testing if we're in a type-only context. Is that correct? If so, don't we have a more general helper for this? If not, how do we know this list is exhaustive?

File: src/Compilers/CSharp/Portable/Parser/LanguageParser_Patterns.cs [hunk 13]
```diff
@@ -35,92 +35,49 @@ private CSharpSyntaxNode CheckRecursivePatternFeature(CSharpSyntaxNode node)
...
-                    // HACK: for error recovery, we prefer a (missing) type.
-                    return this.ParseType(ParseTypeMode.AfterIs);
-                default:
-                    // attempt to disambiguate.
-                    break;
-            }
+                ConstantPatternSyntax cp when ConvertExpressionToType(cp.Expression, out NameSyntax type) => type,
+                TypePatternSyntax tp => tp.Type,
+                DiscardPatternSyntax dp => _syntaxFactory.IdentifierName(ConvertToIdentifier(dp.UnderscoreToken)),
+                var p => p,
```
Comment (agocke) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser_Patterns.cs]: `_ => pattern`?

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs (hunk 12 above)
Comment (gafter) [reply src/Compilers/CSharp/Portable/Parser/LanguageParser.cs]: No, this is checking whether or not a token is named `var` and in a syntactic context such that *if* it were interpreted as a type, it could be the lazily evaluated type `var`. It is mainly trying to rule out contexts in which it is definitely not the `var` type.

---
In reply to: [354968841](https://github.com/dotnet/roslyn/pull/39913#discussion_r354968841) [](ancestors = 354968841)

File: src/Compilers/CSharp/Test/Syntax/Parsing/NullableParsingTests.cs (hunk 11 above)
Comment (gafter) [reply src/Compilers/CSharp/Test/Syntax/Parsing/NullableParsingTests.cs]: The change here is caused by T[]? previously being treated as the start of an expression starting with an index operator.  Now we disambiguate it as a type pattern.  Either way, the `?` isn't part of the pattern, but is an unexpected token that results in a syntax error.  Both interpretations result in a syntax error, so we prefer the type interpretation (which is the change)

---
In reply to: [354611441](https://github.com/dotnet/roslyn/pull/39913#discussion_r354611441) [](ancestors = 354611441)

File: src/Compilers/CSharp/Portable/Parser/LanguageParser_Patterns.cs (hunk 13 above)
Comment (gafter) [reply src/Compilers/CSharp/Portable/Parser/LanguageParser_Patterns.cs]: I prefer this form, so that the result is entirely a function of the already computed input.

---
In reply to: [354971890](https://github.com/dotnet/roslyn/pull/39913#discussion_r354971890) [](ancestors = 354971890)

File: src/Compilers/CSharp/Portable/Binder/Binder_Patterns.cs [hunk 14]
```diff
@@ -99,47 +99,72 @@ private BoundExpression BindSwitchExpression(SwitchExpressionSyntax node, Diagno
...
+            // fall back to a type pattern if it fails as a constant pattern.
+            if (!wasExpression && !IsUnderscore(innerExpression))
+            {
+                if (TryBindAsType(innerExpression, out DiagnosticBag bindAsTypeDiagnostics, out BoundTypeExpression boundType))
+                {
+                    diagnostics.AddRangeAndFree(bindAsTypeDiagnostics);
+                    constantDiagnostics.Free();
+                    hasErrors |= CheckValidPatternType(innerExpression, inputType, boundType.Type, patternTypeWasInSource: true, diagnostics: diagnostics);
+                    CheckFeatureAvailability(innerExpression, MessageID.IDS_FeatureExpressionVariablesInQueriesAndInitializers, diagnostics);
+                    return new BoundTypePattern(node, boundType, inputType, hasErrors);
```
[the remark this answers is not in the document]
Comment (gafter) [reply src/Compilers/CSharp/Portable/Binder/Binder_Patterns.cs]: Sure, but I don't think you'll like it better.

---
In reply to: [354607869](https://github.com/dotnet/roslyn/pull/39913#discussion_r354607869) [](ancestors = 354607869)

**Comment by gafter** [comment]
@agocke I think I've responded to all of your comments.  Do you have any others?

File: src/Compilers/CSharp/Portable/Binder/Binder_Patterns.cs (hunk 14 above)
[the remark this answers is not in the document]
Comment (agocke) [reply src/Compilers/CSharp/Portable/Binder/Binder_Patterns.cs]: I think it's a little bit clearer on the freeing, and a little less clear on behavior. I don't feel strongly either way.

**Comment by jasonmalinowski** [comment]
@gafter: this is now tagging a lot of commits for unrelated things, is this is just a git mistake or something that somebody needs to review?

**Comment by gafter** [comment]
oops.

**Comment by RikkiGibson** [comment]
Looks like GitHub is reporting that roslyn-integration-CI failed, but it actually passed.
